udffsck/udffsck.c

```diff
@@ -0,0 +1,3149 @@
+        dbg("Going to unmap area, ptr: %p\n", ptr);
+#endif
+        munmap(*ptr, size);
+        ptr = NULL;
+        dbg("\tArea unmapped\n");
+    } else {
+        dbg("\tArea is already unmapped\n");
+    }
+}
+
+void map_raw(int fd, uint8_t **ptr, uint64_t offset, size_t size, size_t st_size) {
+    if(*ptr != NULL) {
+        dbg("\tArea is already mapped.\n");
+        return;
+    }
+
+    dbg("\tSize: 0x%x, Alloc size 0x%x\n", st_size, size);
+
+    int prot = PROT_READ;
+    // If is there some request for corrections, we need read/write access to medium
+    if(interactive || autofix) {
+        prot |= PROT_WRITE;
+        dbg("\tRW\n");
+    }
+
+    *ptr = (uint8_t *)mmap(NULL, size, prot, MAP_SHARED, fd, offset);
+    if(ptr == MAP_FAILED) {
+        switch(errno) {
+            case EACCES: dbg("EACCES\n"); break;
+            case EAGAIN: dbg("EAGAIN\n"); break;
+            case EBADF: dbg("EBADF\n"); break;
+            case EINVAL: dbg("EINVAL\n"); break;
+            case ENFILE: dbg("ENFILE\n"); break;
+            case ENODEV: dbg("ENODEV\n"); break;
+            case ENOMEM: dbg("ENOMEM\n"); break;
+            case EPERM: dbg("EPERM\n"); break;
+            case ETXTBSY: dbg("ETXTBSY\n"); break;
+            case EOVERFLOW: dbg("EOVERFLOW\n"); break;
+            default: dbg("EUnknown\n"); break;
+        }
+
+        fatal("\tError maping: %s.\n", strerror(errno));
+        exit(8); 
+    }
+#ifdef MEMTRACE
+    dbg("\tArea allocated, pointer: %p, offset 0x%x\n", ptr, offset);
+#else
+    dbg("\tArea allocated\n");
+#endif
+}
+
+/**
+ * \brief UDF VRS detection function
+ *
+ * This function is trying to find VRS at sector 16.
+ * It also do first attempt to guess sectorsize.
+ *
+ * \param[in] *dev memory mapped device array
+ * \param[in,out] *sectorsize found sectorsize candidate
+ * \param[in] force_sectorsize if -B param is used, this flag should be set
+ *                                and sectorsize should be used automatically.
+ * \return 0 -- UDF succesfully detected, sectorsize candidate found
+ * \return -1 -- found BOOT2 or CDW02. Unsupported for now
+ * \return 1 -- UDF not detected 
+ */
+int is_udf(int fd, uint8_t **dev, int *sectorsize, size_t st_size, int force_sectorsize) {
+    struct volStructDesc vsd;
+    struct beginningExtendedAreaDesc bea;
+    struct volStructDesc nsr;
+    struct terminatingExtendedAreaDesc tea;
+    int ssize = 512;
+    int notFound = 0;
+    int foundBEA = 0;
+    uint32_t chunk = 0;
+    uint32_t chunksize = CHUNK_SIZE;
+
+    //void map_chunk(int fd, uint8_t **dev, uint32_t chunk, uint64_t st_size) {
+
+    for(int it=0; it<5; it++, ssize *= 2) {
+        if(force_sectorsize) {
+            ssize = *sectorsize;
+            it = INT_MAX - 1; //End after this iteration
+            dbg("Forced sectorsize\n");
+        }
+
+        dbg("Try sectorsize %d\n", ssize);
+
+        for(int i = 0; i<6; i++) {
+            chunk = (16*BLOCK_SIZE+i*ssize)/chunksize; 
```

Comment:
This is not correct for disks with UDF blocksize > 2048. VRS always needs to start at byte offset 32768 (16*2048), but size of one VSD in VRS is min(2048, blocksize).

Reply:
If I understood you correctly, fix should be here: 9aa0625

Reply:
You are setting:
```
        if(ssize > BLOCK_SIZE) {
            ssize = BLOCK_SIZE;
        }
```
and then
```
*sectorsize = ssize;
```

It means that sectorsize is set maximally to BLOCK_SIZE? Does not seems to be correct.

I think correct should be:
```
chunk = (16*BLOCK_SIZE+i*min(BLOCK_SIZE,ssize))/chunksize;
```
Code in linux kernel which handles this logic is there: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/fs/udf/super.c#n701

Reply:
You are correct, that is wrong (interesting is all tests passed).
Anyway, I implemented your solution, result is here: f65fc15

Reply:
Maybe it needs better tests? :-)

I think problem is in this part of code:
```
             } else if(vsd.stdIdent[0] == '\0') {
                 if(foundBEA) {
                     continue;
                 }
                 notFound = 1;
                 break;
             } else {
```

When VSD identifier is empty, but BEA was previously found then it skip reading VRS to next block. But empty (unrecorded sector) or unknown VSD identifier means end of VRS (ECMA-167 2/8.3.1).

Therefore this logic needs to be fixed. I bet that above empty identifier caused skip to next sector of size 2048 and because block of size 4096 contains two 2048 sectors, just first was skipped and second correctly loaded.

For inspiration you can look at VRS code which is in udfinfo: https://github.com/pali/udftools/blob/master/udfinfo/readdisc.c#L76

Reply:
Added to TODO.